File: FBSimulatorControl/Management/FBSimulatorControl+PrincipalClass.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorControl+PrincipalClass.h"

#import <objc/runtime.h>

#import <Foundation/Foundation.h>

#import <CoreSimulator/SimServiceContext.h>
#import <CoreSimulator/SimDeviceSet.h>

#import <FBControlCore/FBControlCore.h>

#import "FBSimulatorConfiguration.h"
#import "FBSimulatorServiceContext.h"
#import "FBSimulatorControlConfiguration.h"
#import "FBSimulatorError.h"
#import "FBSimulatorSet.h"
#import "FBSimulatorControlFrameworkLoader.h"

@implementation FBSimulatorControl

#pragma mark Initializers

+ (void)initialize
{
  [FBSimulatorControlFrameworkLoader.essentialFrameworks loadPrivateFrameworksOrAbort];
}

+ (nullable instancetype)withConfiguration:(FBSimulatorControlConfiguration *)configuration error:(NSError **)error
{
  NSError *innerError = nil;
  FBSimulatorServiceContext *serviceContext = [FBSimulatorServiceContext sharedServiceContextWithLogger:configuration.logger];
  SimDeviceSet *deviceSet = [serviceContext createDeviceSetWithConfiguration:configuration error:&innerError];
  if (!deviceSet) {
    return [FBSimulatorError failWithError:innerError errorOut:error];
  }
  FBSimulatorSet *set = [FBSimulatorSet setWithConfiguration:configuration deviceSet:deviceSet delegate:nil logger:[configuration.logger withName:@"simulator_set"] reporter:configuration.reporter error:&innerError];
  if (!set) {
    return [FBSimulatorError failWithError:innerError errorOut:error];
  }
  return [[FBSimulatorControl alloc] initWithConfiguration:configuration serviceContext:serviceContext set:set logger:configuration.logger];
}

- (nullable instancetype)initWithConfiguration:(FBSimulatorControlConfiguration *)configuration serviceContext:(nullable FBSimulatorServiceContext *)serviceContext set:(FBSimulatorSet *)set logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _configuration = configuration;
  _serviceContext = serviceContext;
  _set = set;

  return self;
}

@end

File: FBSimulatorControl/Management/FBSimulatorControl+PrincipalClass.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

@objc(FBSimulatorControl)
public final class FBSimulatorControl: NSObject {

  // MARK: - Properties

  @objc public var configuration: FBSimulatorControlConfiguration
  @objc public let serviceContext: FBSimulatorServiceContext
  @objc public let set: FBSimulatorSet

  // MARK: - Initializers

  @objc(withConfiguration:error:)
  public class func withConfiguration(_ configuration: FBSimulatorControlConfiguration) throws -> FBSimulatorControl {
    FBSimulatorControlFrameworkLoader.essentialFrameworks.loadPrivateFrameworksOrAbort()
    let serviceContext = try FBSimulatorServiceContext.sharedServiceContext(withLogger: configuration.logger)
    let deviceSet = try serviceContext.createDeviceSet(with: configuration)
    let set = FBSimulatorSet.set(
      withConfiguration: configuration,
      deviceSet: deviceSet,
      delegate: nil,
      logger: configuration.logger.withName("simulator_set"),
      reporter: configuration.reporter)
    return FBSimulatorControl(configuration: configuration, serviceContext: serviceContext, set: set)
  }

  /**
   Fork addition: bootstraps with an injected Xcode developer directory.

   Sandboxed hosts cannot resolve the developer directory via `xcode-select`,
   so they pass the directory obtained from a security-scoped bookmark here
   before any Xcode-path-dependent code runs.
   */
  @objc(withConfiguration:developerDirectory:error:)
  public class func withConfiguration(_ configuration: FBSimulatorControlConfiguration, developerDirectory: String?) throws -> FBSimulatorControl {
    FBXcodeConfiguration.setInjectedDeveloperDirectory(developerDirectory)
    return try withConfiguration(configuration)
  }

  private init(configuration: FBSimulatorControlConfiguration, serviceContext: FBSimulatorServiceContext, set: FBSimulatorSet) {
    self.configuration = configuration
    self.serviceContext = serviceContext
    self.set = set
    super.init()
  }
}
